File: python/twitter/utils/postgres_utils.py
import os
from sqlalchemy.sql import text
from sqlalchemy import create_engine
import logging

# def connectToPostgres():
#     url = getUrl()
#     engine = create_engine(url)
#     connection = engine.connect()
#     return connection

_connections = []


def get_connection():
    if _connections:
        return _connections[0]
    else:
        connection = create_engine(get_connect_str()).connect()
        _connections.append(connection)
        return connection


def execute_query(sql):
    connection = get_connection()
    cursor = connection.execute(text(sql).execution_options(autocommit=True))
    return cursor


def get_connect_str():
    url = os.environ['POSTGRES_URL']
    user = os.environ['POSTGRES_USER']
    password = os.environ['POSTGRES_PASSWORD']
    return "postgresql+psycopg2://{0}:{1}@localhost:5432/postgres".format(user, password)


def insert_sentiment(user, sentiment, score, id, col1, col2):

    log = logging.getLogger("app." + __name__)
    connection = get_connection()
    sql = "UPDATE tweets_from_{0} \
            SET {1} = {2}, {3} = '{4}' \
            WHERE id = {5}"\
        .format(user, col1, score, col2, sentiment, id)
    connection.execute(text(sql).execution_options(autocommit=True))
    log.info("Tweet for user: {}, with id: {}, with score: {}, and sentiment: "
             "{} was successfully inserted".format(user, id, score, sentiment))



# def insert_clean_text(text, id):
#     connection = get_connection()
#     if text.find("'"):
#         text = text.replace("'", "''")
#     sql = "UPDATE replies_from_dm_me_your_cats_friends \
#             SET clean_after_python_text = '{0}' WHERE id = {1}".format(text, id)
#     print(sql)
#     connection.execute(text(sql).execution_options(autocommit=True))


def insert_clsf_sentiment(user, score, id, col1, col2):

    log = logging.getLogger("app." + __name__)
    sentiment = "positive" if score == 1 else "negative"
    connection = get_connection()
    sql = "UPDATE tweets_from_{0} \
            SET {1} = '{2}' ," \
           "{3} = {4} \
            WHERE id = {5}" \
        .format(user, col1, sentiment, col2, score, id)
    connection.execute(text(sql).execution_options(autocommit=True))
    log.info("Tweet for user: {}, with id: {}, with score: {}, and sentiment: "
             "{} was successfully inserted".format(user, id, score, sentiment))


def get_ids_text(user):
    connection = get_connection()
    sql = "SELECT id, clean_text FROM tweets_from_{0}".format(user)
    cursor = connection.execute(text(sql)
                                .execution_options(autocommit=True))
    return cursor

File: java/src/main/java/com.mazhara.utils/postresql/PostgresUtils.java
package com.mazhara.utils.postresql;


import com.mazhara.utils.preprocess.TextUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import twitter4j.Status;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class PostgresUtils {

    static Logger LOGGER = LoggerFactory.getLogger(PostgresUtils.class);


    public static List<String> getCredentials(){

        final String urlDB = System.getenv("POSTGRES_URL");
        final String userDB = System.getenv("POSTGRES_USER");
        final String passwordDB = System.getenv("POSTGRES_PASSWORD");
        return  new ArrayList() {{
            add(urlDB);
            add(userDB);
            add(passwordDB);
        }};
    }


    public static Connection connectToPostgres()  {
        Connection conn = null;
        try {
            List<String> creds = getCredentials();
            conn = DriverManager.getConnection(creds.get(0), creds.get(1), creds.get(2));
            LOGGER.info("Connected to the PostgreSQL server successfully.");
        } catch (SQLException e) {
            System.out.println(e.getMessage());
        }
        return conn;
    }


    //TODO: change score col type to show double values
    public static void createTable(Connection conn, String user){
        try{
            String tweets = String.format("DROP TABLE IF EXISTS tweets_from_%s; CREATE TABLE  " +
                    "tweets_from_%s " +
                    "(id NUMERIC NOT NULL PRIMARY KEY, text varchar(280) NOT NULL," +
                    "clean_text varchar(280), CoreNLP_mark int, CoreNLP_sentiment varchar(20)," +
                    "vader_mark int, vader_sentiment varchar(20)," +
                    "svmclassifier_mark int, svmclassifier_sentiment varchar(20), " +
                    "mbclassifier_mark int, mbclassifier_sentiment varchar(20), " +
                    "lstm_mark int, lstm_sentiment varchar(20));",user, user);

            PreparedStatement ps = conn.prepareStatement(tweets);
            ps.executeUpdate();
            ps.close();
            LOGGER.info("Table is created for user :" + user);
        }catch (SQLException e) {
            System.out.println(e.getMessage());
        }
    }


    public static void insertTweetsToTable(Connection conn, Status status, String userSource){

        String text = status.getText();
        String userName = status.getUser().getScreenName();
        try{
            if (text.contains("\'") == true){
                text = text.replace("\'", "\''");
            }
            if (userName.contains("\'") == true){
                userName = userName.replace("\'", "\''");
            }

            String insertTweets = String.format("INSERT INTO  " +
                    "tweets_from_%s (id, text) VALUES (%d, '%s') " +
                    "ON CONFLICT DO NOTHING;",userSource, status.getId(), text);

            PreparedStatement ps = conn.prepareStatement(insertTweets);
            ps.executeUpdate();
            ps.close();
        }catch (SQLException e) {
            System.out.println(e.getMessage());
        }
        LOGGER.info("Inserted record for user " + userName);

    }

    public static void cleanTextField(Connection conn, String user){

        String getTweets = String.format("SELECT * FROM " +
                "tweets_from_%s",user);
        try {
            PreparedStatement ps = conn.prepareStatement(getTweets);
            ResultSet resultSet = ps.executeQuery();
            while (resultSet.next()) {

                String tweet = resultSet.getString("text");
                tweet = TextUtils.cleanText(tweet);
                long id = resultSet.getLong("id");
                String updateTweet = String.format("UPDATE   " +
                        "tweets_from_%s SET clean_text = '%s' " +
                        "WHERE id = %d;",user, tweet, id);
                ps = conn.prepareStatement(updateTweet);
                ps.executeUpdate();

            }
        }catch (SQLException e) {
            System.out.println(e.getMessage());
        }
        LOGGER.info("All tweets have been cleaned!");

    }

    public static ResultSet getAllTweets(Connection conn, String user){

        String getTweets = String.format("SELECT * FROM " +
                "tweets_from_%s;",user);
        ResultSet resultSet = null;
        try {
            PreparedStatement ps = conn.prepareStatement(getTweets);
            resultSet = ps.executeQuery();
        }catch (SQLException e) {
            System.out.println(e.getMessage());
        }
        LOGGER.info("All tweets have been loaded!");
        return resultSet;
    }


    public static void insertSentiment(Connection conn, String user, int sentimentScore, String sentimentType, long tweetId){

        LOGGER.info("user " + user);
        LOGGER.info(String.format("tweets_from_%s", user));
        LOGGER.info(String.format("where id = %d", tweetId));


        String getTweet = String.format("UPDATE  " +
                "tweets_from_%s SET CoreNLP_mark = %d ," +
                "CoreNLP_sentiment = '%s' where id = %d;", user, sentimentScore, sentimentType, tweetId);
        try {
            PreparedStatement ps = conn.prepareStatement(getTweet);
            ps.executeUpdate();
            ps.close();
        }catch (SQLException e) {
            System.out.println(e.getMessage());
        }
        LOGGER.info("Inserted the sentiment mark ...");
    }


}
